Monads.NET/Maybe.cs

```diff
@@ -98,10 +99,44 @@ public static TInput IfDo<TInput>(this TInput o, Func<TInput, bool> evaluator, A
         public static TResult With<TKey, TResult>(this IDictionary<TKey, TResult> o, TKey key)
             where TResult : class
         {
-            return (o == null || !o.ContainsKey(key)) ? null : o[key];
+            TResult value;
+            return (o == null || !o.TryGetValue(key, out value)) ? null : value;
         }
 
         /// <summary>
+        /// Allows for a null safe accessing of an item
+        /// </summary>
+        /// <typeparam name="TInput">Reference type of object being extended</typeparam>
+        /// <typeparam name="TResult">Reference type of object being returned</typeparam>
+        /// <param name="o">instance of object being extended</param>
+        /// <param name="evaluator">function that acts on object to return a result</param>
+        /// <returns>null if object is null or an instance of TResult</returns>
+        public static Task<TResult> WithAsync<TInput, TResult>(this TInput o, Func<TInput, Task<TResult>> evaluator)
+            where TResult : class
+            where TInput : class
+        {
+            return (o == null) ? TaskEx.FromResult<TResult>(null) : evaluator(o);
```

Comment:
There a reason TaskEx is used here? `Task.FromResult` is perfectly acceptable and has broader support.

Also `where TInput : class` is unnecessary, right?